File: replication/src/spawn.rs
//! `SPAWN`: a node coming into existence on the far side.
//!
//! # Layout, measured over 112 captured packets
//!
//! ```text
//! [u8 0x04][u8 scene][u32 spawner][u32 net_id][u32 sync_count][u32 name_len]
//! [u32 sync_net_id x sync_count][name_len bytes of name, NUL included][state]
//! ```
//!
//! All 112 captured spawns fit this layout, and each field was pinned by
//! correlation:
//!
//! - `scene` indexes the spawner's `_spawnable_scenes` (`level.tscn:75`,
//!   three entries): 0 is a player, 1 a robot, 2 a bullet.
//! - `spawner` is 1 in every packet, which is the path-cache id
//!   `SIMPLIFY_PATH` gave `main/Level/MultiplayerSpawner`.
//! - `sync_count` synchronizer ids follow, and they are always
//!   `net_id + 1 ..= net_id + sync_count`: the node and each synchronizer
//!   under it get consecutive ids.
//!
//! # A correction worth keeping
//!
//! The first version of this module read a fixed five-word header and called
//! the third word a "kind", because in 111 packets it was 1 and the fifth
//! word was always `net_id + 1`. The 112th had a 2 there, a two-byte "name"
//! that was not a string, and a state that would not decode, so it was stored
//! unparsed as `Spawn::Raw`.
//!
//! That packet is the host's own player, `"1"`. It is the only spawned node
//! with two synchronizers the server owns: `ServerSynchronizer` and
//! `InputSynchronizer`. The two-byte non-name was its second synchronizer id,
//! `0b 00 00 00`, read half-way through. Nothing about it was unusual. The
//! misreading only surfaced when this crate had to *write* a spawn and needed
//! to know what that word meant.
//!
//! # The state is compact Variants, and the sizes prove it
//!
//! Every `spawn = true` property of every listed synchronizer, in order, in
//! the compact encoding SYNC uses:
//!
//! ```text
//! robot        transform 52 + health 2 + state 2 + target 16 + dead 1   =  73
//! bullet       transform 52                                            =  52
//! client       transform 52 + id 5 + model 52 + motion 12 + anim 2      = 123
//! host "1"     the same with id 2 (= 120), plus input 16x3 + 12 + 1 + 1 = 182
//! ```
//!
//! The client's player lists one synchronizer and the host's lists two
//! because `level.gd` gives each player's `InputSynchronizer` to that
//! player's peer: the server includes only the synchronizers it owns.

/// Why a spawn packet could not be read.
#[derive(Debug, PartialEq, Eq, Clone)]
pub enum SpawnError {
    /// The first byte was not [`COMMAND_SPAWN`].
    WrongCommand {
        /// What was found.
        found: u8,
    },
    /// The packet ended inside a field.
    Truncated,
    /// The name was not NUL-terminated within its stated length.
    Unterminated,
}

/// `SPAWN`.
pub const COMMAND_SPAWN: u8 = 0x04;

/// A spawn packet.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Spawn {
    /// Index into the spawner's `_spawnable_scenes`.
    pub scene: u8,
    /// The spawner's path-cache id.
    pub spawner: u32,
    /// The id the spawned node will be addressed by.
    pub net_id: u32,
    /// Ids of the synchronizers under it that the sender owns, in order.
    pub sync_ids: Vec<u32>,
    /// The node's name, without its NUL.
    pub name: String,
    /// The `spawn = true` properties of those synchronizers, compact-encoded
    /// in order.
    ///
    /// Left as bytes: decoding them needs each synchronizer's field list,
    /// which the receiver has and the packet does not.
    pub state: Vec<u8>,
}

fn word(packet: &[u8], at: usize) -> Result<u32, SpawnError> {
    packet
        .get(at..at + 4)
        .and_then(|s| s.try_into().ok())
        .map(u32::from_le_bytes)
        .ok_or(SpawnError::Truncated)
}

impl Spawn {
    /// Reads one `SPAWN` packet.
    ///
    /// # Errors
    ///
    /// Any of [`SpawnError`].
    pub fn parse(packet: &[u8]) -> Result<Self, SpawnError> {
        match packet.first() {
            Some(&COMMAND_SPAWN) => {}
            Some(&found) => return Err(SpawnError::WrongCommand { found }),
            None => return Err(SpawnError::Truncated),
        }
        let scene = *packet.get(1).ok_or(SpawnError::Truncated)?;
        let spawner = word(packet, 2)?;
        let net_id = word(packet, 6)?;
        let sync_count = word(packet, 10)? as usize;
        let name_len = word(packet, 14)? as usize;
        let mut at = 18;
        let mut sync_ids = Vec::with_capacity(sync_count.min(64));
        for _ in 0..sync_count {
            sync_ids.push(word(packet, at)?);
            at += 4;
        }
        let name = packet.get(at..at + name_len).ok_or(SpawnError::Truncated)?;
        let name = name.strip_suffix(&[0]).ok_or(SpawnError::Unterminated)?;
        at += name_len;
        Ok(Self {
            scene,
            spawner,
            net_id,
            sync_ids,
            name: String::from_utf8_lossy(name).into_owned(),
            state: packet[at..].to_vec(),
        })
    }

    /// Writes it back out.
    #[must_use]
    pub fn encode(&self) -> Vec<u8> {
        let mut out =
            Vec::with_capacity(19 + 4 * self.sync_ids.len() + self.name.len() + self.state.len());
        out.push(COMMAND_SPAWN);
        out.push(self.scene);
        out.extend_from_slice(&self.spawner.to_le_bytes());
        out.extend_from_slice(&self.net_id.to_le_bytes());
        let count = u32::try_from(self.sync_ids.len()).unwrap_or(u32::MAX);
        out.extend_from_slice(&count.to_le_bytes());
        let name_len = u32::try_from(self.name.len() + 1).unwrap_or(u32::MAX);
        out.extend_from_slice(&name_len.to_le_bytes());
        for id in &self.sync_ids {
            out.extend_from_slice(&id.to_le_bytes());
        }
        out.extend_from_slice(self.name.as_bytes());
        out.push(0);
        out.extend_from_slice(&self.state);
        out
    }
}

File: dart/lib/src/spawn.dart
// SPDX-License-Identifier: Apache-2.0
//
// `SPAWN`: a node coming into existence on the far side (plan.md DR-7, DR-7a).
//
// ## Layout, measured over 112 captured packets
//
//     [u8 0x04][u8 scene][u32 spawner][u32 net_id][u32 sync_count]
//     [u32 name_len][u32 sync_net_id x sync_count][name, NUL included][state]
//
// - `scene` indexes the spawner's `_spawnable_scenes` (`level.tscn:75`, three
//   entries): 0 is a player, 1 a robot, 2 a bullet.
// - `spawner` is the path cache id of `main/Level/MultiplayerSpawner`, which
//   is why that path has to be announced before any spawn that names it.
// - the synchronizer ids are always `net_id + 1 ..= net_id + sync_count`: the
//   node and each synchronizer under it take consecutive ids.
//
// ## The third word is a count, not a kind
//
// Worth repeating here because the Rust side got it wrong first and the
// mistake was invisible: in 111 of 112 packets that word is 1, and the 112th
// is the host's own player `"1"` -- the only spawned node whose server owns
// two synchronizers, `ServerSynchronizer` and `InputSynchronizer`. Reading it
// as a kind made that packet's second synchronizer id look like a two-byte
// name and its state undecodable. Nothing about the packet was unusual; the
// reading was.
//
// ## The state is compact Variants
//
// Every `spawn = true` property of every listed synchronizer, in order, in the
// same encoding SYNC uses. A bullet's is one Transform3D; a robot's is the
// five its scene lists, `health` and `dead` among them -- which is how those
// two cross exactly once and never stream.

import 'dart:convert';
import 'dart:typed_data';

import 'sync.dart';
import 'variant.dart';

/// The `SceneMultiplayer` command byte for a spawn.
const int commandSpawn = 0x04;

/// Fixed words before the synchronizer ids: command, scene, spawner, net id,
/// sync count, name length.
const int spawnHeaderLength = 18;

/// A node arriving.
class Spawn {
  const Spawn({
    required this.scene,
    required this.spawner,
    required this.netId,
    required this.syncIds,
    required this.name,
    required this.state,
    this.stateBytes,
  });

  /// A spawn whose state has already been written, by the same sink the
  /// streamed fields go through. See [SyncRecord.encoded].
  const Spawn.encoded({
    required this.scene,
    required this.spawner,
    required this.netId,
    required this.syncIds,
    required this.name,
    required Uint8List this.stateBytes,
  }) : state = const <SyncField>[];

  /// Index into the spawner's spawnable scene list.
  final int scene;

  /// The path cache id of the MultiplayerSpawner doing the spawning.
  final int spawner;

  /// The object id. DESPAWN names this one.
  final int netId;

  /// One per synchronizer the sender owns on this node.
  final List<int> syncIds;

  /// The node's name, which is its identity in every path built from it.
  final String name;

  /// The `spawn = true` properties, in order, compact. Empty on a spawn built
  /// from bytes.
  final List<SyncField> state;

  /// The encoded state, when this spawn was built from one.
  final Uint8List? stateBytes;

  static Spawn parse(final Uint8List packet) {
    if (packet.isEmpty || packet[0] != commandSpawn) {
      throw const FormatException('not a SPAWN packet');
    }
    if (packet.length < spawnHeaderLength) {
      throw const FormatException('SPAWN is cut short');
    }
    final ByteData view = ByteData.sublistView(packet);
    final int scene = packet[1];
    final int spawner = view.getUint32(2, Endian.little);
    final int netId = view.getUint32(6, Endian.little);
    final int syncCount = view.getUint32(10, Endian.little);
    final int nameLength = view.getUint32(14, Endian.little);

    final int idsEnd = spawnHeaderLength + 4 * syncCount;
    if (idsEnd + nameLength > packet.length) {
      throw FormatException(
        'SPAWN claims $syncCount ids and a $nameLength byte name, '
        '${packet.length - spawnHeaderLength} bytes left',
      );
    }
    final List<int> syncIds = <int>[
      for (int i = 0; i < syncCount; i++)
        view.getUint32(spawnHeaderLength + 4 * i, Endian.little),
    ];
    if (nameLength == 0) {
      throw const FormatException('SPAWN has no name');
    }
    // The length counts the NUL.
    final String name =
        utf8.decode(packet.sublist(idsEnd, idsEnd + nameLength - 1));

    final List<SyncField> state = <SyncField>[];
    final int stateAt = idsEnd + nameLength;
    int at = stateAt;
    while (at < packet.length) {
      final DecodedVariant read = decodeCompactVariant(packet, at);
      state.add((type: read.type, value: read.value));
      at = read.next;
    }
    return Spawn(
      scene: scene,
      spawner: spawner,
      netId: netId,
      syncIds: syncIds,
      name: name,
      state: state,
      // Kept as well as decoded: applying a spawn means handing these bytes to
      // the same reader a streamed record goes through, and writing the values
      // back out to get them would be work done twice.
      stateBytes: Uint8List.sublistView(packet, stateAt),
    );
  }

  Uint8List encode() {
    final Uint8List nameBytes = utf8.encode(name);
    final ByteData head = ByteData(spawnHeaderLength)
      ..setUint8(0, commandSpawn)
      ..setUint8(1, scene)
      ..setUint32(2, spawner, Endian.little)
      ..setUint32(6, netId, Endian.little)
      ..setUint32(10, syncIds.length, Endian.little)
      ..setUint32(14, nameBytes.length + 1, Endian.little);

    final BytesBuilder out = BytesBuilder(copy: false)
      ..add(head.buffer.asUint8List());
    for (final int id in syncIds) {
      final ByteData word = ByteData(4)..setUint32(0, id, Endian.little);
      out.add(word.buffer.asUint8List());
    }
    out
      ..add(nameBytes)
      ..addByte(0);
    if (stateBytes != null) {
      out.add(stateBytes!);
    } else {
      for (final SyncField field in state) {
        out.add(encodeCompactField(field.type, field.value));
      }
    }
    return out.takeBytes();
  }
}
